### ISSUE TITLE
Bug: [require-await] will auto fix into failing typescript code if the function is typed as returning a promise

### ISSUE THREAD

**Issue by beaussan** [typescript-eslint/typescript-eslint#10084]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=PTAEBUE8AcFMGdQGMD2BbNBLAdgc1AGYBO6oALgBZGYAmo0AhkWZKADaYBGAUC3KAFlIAMWxR%2BAXlAAKAJSgJAPlAAFElniwAPNgCuaTrCKLu3VNnhlQNWGwasp0tPABcoPQaOy3a9Jk1aAG4otMpKoADe3KCgRLBkukTY7rAA7qrq-rDScfAobIGwCsqaZOCYaLAoumQ5CPmFADSgzrKyANzcAL6dpiCgAMoU1Wx0hqAoANZmKBZWBLrYSGSYswDqmJQAgvCQSwDCDGxsbkKi4kVSDLtLMvLhUTEMqQyb1rb20gCMAAw-HdFYvFEskvp0un0wEMRnQjnlQOMps1ODVCK82PBmgxsHQCJgAB4IBFxBiTRCUIp8BBIajQKyoGwzOaERbLVbYDaUao7PZIQ7HABCNQAkmIYAhTiIxZJQNdeXdipFAXEEklQGDupDQABxIj2WUk5pkWBICjYTBII5sVhvFUgxAMeiZTTI1EUuKgfzuFCyl6bbj9cAUL0Ua4I2CwZJoVipFBESZMao4whx8jBxBEXRsIoEdE4XAAOgDYGFBDTDDIAHJELB8XBlrA6LH44nFjRmm94MMs2MijQUEh9JHjTQmZZQF2YcJ0UKyBtjgA5FBkSXncUKWU3JAKh7K4Fq3waWAF3INbJfAEQoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEI0QR1gEtpEBaAQwHdzCAXdKAewDNnIAacbSAAVoE8ADigDG0QoNqkU8QgDtaAehIFiZKjXoZIiaNEbRIXAL4hjQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// Types comming from thrid party lib
type MyFnType = () => Promise<number>

const delay = (ms: number): Promise<void> => {
  return new Promise(resolve => setTimeout(resolve, ms));
};


// Should be ok
const functionWithAsyncCall: MyFnType = async () => {
  await delay(100);
  return 1;
}

// Should also be ok, but fails, and fixes breaks the typescript code
const functionWithouAsyncCallButInTypes: MyFnType = async () => {
  return 1;
}

// Gray area, technically it returns a promise, but there is no await
// This has been my workaround for this rule failing.
// If that's expected workaround, it should be documented
const shouldFailButWillNot: MyFnType = async () => {
  return Promise.resolve(1);
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "require-await": "off",
    "@typescript-eslint/require-await": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I was expecting the rule not to fail since the types are explicitly requesting for a promise

### Actual Result

Fails because there is async, and the fix applied is to remove the async, but then that fails the typescript compilation, making a eslint --fix produce broken code

**Comment by beaussan** [comment]
Good to know!

So the only way to silence it is via a eslint disable directive in a case where the public api forces you do do a async function, but you have no await ? Can't the lint rule see that the forced type (and so not the one inferred) is explicitly requesting a async like function ? 

Because, in my mental model:
```ts
// Types comming from thrid party lib
type MyFnType = () => Promise<number>

const delay = (ms: number): Promise<void> => {
  return new Promise(resolve => setTimeout(resolve, ms));
};


// Should be ok, but fails
const functionWithouAsyncCallButInTypes: MyFnType = async () => {
  return 1;
}

// Should fails since it's doing a async without await and there is no forced type for it
const functionWithouAsyncAndWithoutCastedType = async () => {
  return 1;
}
```

**Comment by bradzacher** [comment]
But who's to say that the type is correct? Just because the type expects a promise doesn't mean that it should expect a promise. I.e. The rule could be flagging that your code is due for a refactor to remove the asynchrony entirely!

In the general case if you're using `async` without a promise then it's a code smell and the asynchrony is usually unnecessary.
In some cases the useless Promise is required by an external API - in which case the fix is to either use a disable comment to indicate that it's intentional, or keep the `async` but explicitly wrap the return in a `Promise.resolve` to similarly mark it as intentional.

**Comment by kirkwaiblinger** [comment]
FWIW - speaking for myself only, I feel like this is one of the rules that's more of an informative aid in development than a "you're probably doing something wrong". [no-await-in-loop](https://eslint.org/docs/latest/rules/no-await-in-loop) is another one like that, where in some cases you might be able to do it a bit better, but in other cases you also might very well need to use `await` in a loop 🤷

**Comment by bradzacher** [comment]
There are definitely valid uses for an `await` in a loop -- but often it's better to use `Promise.all(array.map(async (value) => { ... }))`. Which is what that rule is trying to get you to do -- parallelise your promises rather than synchronously evaluate them. But there are still valid usecases for synchronously processing promises!

This rule is the same yeah -- most of the usecases for `async` without `await` are smelly -- but there are some valid cases such as conforming to an external API.
